examples/analysis/IDE.flix

```diff
@@ -1,182 +1,210 @@
-        pub def hash(x: Value): Int32 = match x {
-            case Top    => 1
-            case Bot    => 2
-            case Cst(n) => n
-        }
+instance JoinLattice[Value] {
+    pub def leastUpperBound(e1: Value, e2: Value): Value = match (e1, e2) {
+        case (Value.Bot, x) => x
+        case (x, Value.Bot) => x
+        case (Value.Cst(n1), Value.Cst(n2)) => if(n1 == n2) Value.Cst(n1) else Value.Top
+        case _ => Value.Top
     }
+}
 
-    instance PartialOrder[Value] {
-        pub def partialCompare(e1: Value, e2: Value): Bool = match (e1, e2) {
-            case (Value.Bot, _) => true
-            case (Value.Cst(n1), Value.Cst(n2)) => n1 == n2
-            case (_, Value.Top) => true
-            case _ => false
-        }
+instance MeetLattice[Value] {
+    pub def greatestLowerBound(e1: Value, e2: Value): Value = match (e1, e2) {
+        case (Value.Top, x) => x
+        case (x, Value.Top) => x
+        case (Value.Cst(n1), Value.Cst(n2)) => if(n1 == n2) Value.Cst(n1) else Value.Bot
+        case _ => Value.Bot
     }
+}
 
-    instance JoinLattice[Value] {
-        pub def leastUpperBound(e1: Value, e2: Value): Value = match (e1, e2) {
-            case (Value.Bot, x) => x
-            case (x, Value.Bot) => x
-            case (Value.Cst(n1), Value.Cst(n2)) => if(n1 == n2) Value.Cst(n1) else Value.Top
-            case _ => Value.Top
-        }
+instance ToString[Value] {
+    pub def toString(x: Value): String = match x {
+        case Top    => "Top"
+        case Cst(n) => "Cst(${n})"
+        case Bot    => "Bot"
     }
+}
 
-    instance MeetLattice[Value] {
-        pub def greatestLowerBound(e1: Value, e2: Value): Value = match (e1, e2) {
-            case (Value.Top, x) => x
-            case (x, Value.Top) => x
-            case (Value.Cst(n1), Value.Cst(n2)) => if(n1 == n2) Value.Cst(n1) else Value.Bot
-            case _ => Value.Bot
-        }
+instance Boxable[Value] {
+    pub def box(x: Value): Boxed = {
+        let value = x as ##java.lang.Object;
+        let compare = (o1, o2) -> Order.compare(o1 as Value, o2 as Value);
+        let toString = o -> ToString.toString(o as Value);
+        BoxedObject(value, compare, toString)
     }
-
-    instance ToString[Value] {
-        pub def toString(x: Value): String = match x {
-            case Top    => "Top"
-            case Cst(n) => "Cst(${n})"
-            case Bot    => "Bot"
-        }
+    pub def unbox(x: Boxed): Value = match x {
+        case BoxedObject(v, _, _) => v as Value
+        case _                    => ?bug
     }
+}
 
-    enum Transformer {
-        case BotTrans,
-        case NonBotTrans(Int,Int,Value)
+instance Order[Value] {
+    pub def compare(x: Value, y: Value): Comparison = match (x, y) {
+        case (Bot, Bot)         => EqualTo
+        case (Cst(n1), Cst(n2)) => n1 <=> n2
+        case (Top, Top)         => EqualTo
+        case (Bot, _)           => LessThan
+        case _                  => GreaterThan
     }
+}
```

Comment:
I know this is just an example, but I found some inconsistencies. The definitions of partial order, lub and glb imply that no total order of Value exist since Cst values have no ordering.

Reply:
It is a good point and the ergonomics are not great.

It does, however, make sense: The partial order defines the lattice (which is used for the semantics of the Datalog program). The total order is unrelated to the partial order and is not used in the semantics. It can, however, be used to print the facts in some order!

Reply:
Oh, that makes sense, but could be misunderstood. Maybe a comment about this should exist?